src/adapter/src/catalog/apply.rs

```diff
@@ -1028,22 +1099,79 @@ fn sort_updates(updates: Vec<StateUpdate>) -> Vec<StateUpdate> {
         }
     }
 
-    // Sort item updates by GlobalId.
+    /// Sort item updates by GlobalId.
     fn sort_item_updates(
         item_updates: Vec<(mz_catalog::durable::Item, StateDiff)>,
-    ) -> Vec<StateUpdate> {
+    ) -> VecDeque<(mz_catalog::durable::Item, StateDiff)> {
         item_updates
             .into_iter()
             .sorted_by_key(|(item, _diff)| item.id)
-            .map(|(item, diff)| StateUpdate {
-                kind: StateUpdateKind::Item(item),
-                diff,
-            })
             .collect()
     }
     let item_retractions = sort_item_updates(item_retractions);
     let item_additions = sort_item_updates(item_additions);
 
+    /// Sort temporary item updates by GlobalId.
+    fn sort_temp_item_updates(
+        temp_item_updates: Vec<(TemporaryItem, StateDiff)>,
+    ) -> VecDeque<(TemporaryItem, StateDiff)> {
+        temp_item_updates
+            .into_iter()
+            .sorted_by_key(|(item, _diff)| item.id)
+            .collect()
+    }
+    let temp_item_retractions = sort_temp_item_updates(temp_item_retractions);
+    let temp_item_additions = sort_temp_item_updates(temp_item_additions);
+
+    /// Merge sorted temporary and non-temp items.
+    fn merge_item_updates(
+        mut item_updates: VecDeque<(mz_catalog::durable::Item, StateDiff)>,
+        mut temp_item_updates: VecDeque<(TemporaryItem, StateDiff)>,
+    ) -> Vec<StateUpdate> {
```

Comment:
Can you remind me why we need to sort updates again?

From what I remember it's because we want to process updates for objects before processing what they depend on?

Reply:
Yeah we sort the updates in dependency order so that we can apply them in dependency order. For example creating a database before creating a schema in that database.

The doc-comments say:
```
/// Sort [`StateUpdate`]s in dependency order.
```

Any suggestions on what to add to make this more obvious?